core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/NamespaceChangeStreamListener.java

```diff
@@ -0,0 +1,245 @@
+    this.callbackQueue = new ArrayDeque<>();
+    this.networkMonitor.addNetworkStateListener(this);
+  }
+
+  @Override
+  public void onNetworkStateChanged() {
+    if (!this.networkMonitor.isConnected()) {
+      this.stop();
+      this.nsConfig.setStale();
+    } else {
+      this.start();
+    }
+  }
+
+  /**
+   * Opens the stream in a background thread.
+   */
+  public void start() {
+    nsLock.writeLock().lock();
+    try {
+      if (streamThread != null) {
+        return;
+      }
+      streamThread =
+          new Thread(new NamespaceChangeStreamRunner(
+              new WeakReference<>(this), networkMonitor, logger));
+      streamThread.start();
+    } finally {
+      nsLock.writeLock().unlock();
+    }
+  }
+
+  /**
+   * Stops the background stream thread.
+   */
+  public void stop() {
+    nsLock.writeLock().lock();
+    try {
+      if (streamThread == null) {
+        return;
+      }
+
+      streamThread.interrupt();
+      try {
+        if (streamThread.isAlive()) {
+          close();
+          streamThread.join();
+        }
+      } catch (final Exception e) {
+        return;
+      }
+      streamThread = null;
+    } catch (Exception e) {
+      e.printStackTrace();
+    } finally {
+      nsLock.writeLock().unlock();
+    }
+  }
+
+  void queueWatcher(final Callback<ChangeEvent<BsonDocument>, Object> callback) {
+    callbackQueue.add(callback);
+  }
+
+  void clearWatchers() {
+    for (int i = 0; i < callbackQueue.size(); i++) {
+      callbackQueue.remove().onComplete(
+          OperationResult.<ChangeEvent<BsonDocument>, Object>failedResultOf(null));
+    }
+  }
+
+  void close() {
+    if (currentStream != null) {
+      try {
+        currentStream.close();
+      } catch (final IOException e) {
+        e.printStackTrace();
+      } finally {
+        currentStream = null;
+      }
+    }
+  }
+
+  /**
+   * Opens the next next.
+   */
+  void stream() {
+    logger.info("stream START");
+    if (!networkMonitor.isConnected()) {
+      logger.info("stream END - Network disconnected");
+      return;
+    }
+    if (!authMonitor.isLoggedIn()) {
+      logger.info("stream END - Logged out");
+      return;
+    }
+
+    if (nsConfig.getStaleDocumentIds().isEmpty()) {
+      logger.info("stream END - No stale documents");
+      return;
+    }
+
+    final Document args = new Document();
+    args.put("database", namespace.getDatabaseName());
+    args.put("collection", namespace.getCollectionName());
+    args.put("ids", nsConfig.getStaleDocumentIds());
+
+    currentStream =
+        service.streamFunction(
+            "watch",
+            Collections.singletonList(args),
+            ChangeEvent.changeEventCoder);
+    try {
+      while (currentStream.isOpen()) {
+        final Event<ChangeEvent<BsonDocument>> event = currentStream.nextEvent();
+        if (event.getEventType() == EventType.ERROR) {
+          throw event.getError();
```

Comment:
Maybe we should just emit this error through the DataSynchronizer?

Reply:
Because of the importance of these errors. I would have maybe something like the error emitter passed into there (may need that ifc back) but you want to make sure that we make that a special emitter that can look at what went wrong. If it's an invalid session or something like that via handling the rich error, that will need to bubble up to the right place. I would make a TODO for this and file a ticket so that we can move on with this PR. Create the ticket and then do `TODO(STITCH-XXX)` on this line replacing the ticket number.